Prevent `systemd` from killing the Upgrade Watcher process after the main Agent process has crashed

ycombinator opened the pull request:
## What does this PR do?

This PR adds the `KillMode=process` setting to Elastic Agent's `systemd` unit configuration file, installed at `/etc/systemd/system/elastic-agent.service`. 

## Why is it important?

The [`KillMode`](https://www.freedesktop.org/software/systemd/man/systemd.kill.html#KillMode=)  setting governs which processes under the service will be killed by `systemd`. The default value of the setting is `control-group`.  With this value, all processes under the service are killed. However, with the `process` value, only the main process under the service is killed while the others are left running.

In the context of Elastic Agent, the main process it the Agent process.  During an ongoing upgrade, there is a second process in the same cgroup as the main process: the Upgrade Watcher process. 

In the scenario when the main process crashes or is not reporting a healthy status repeatedly, we need the Upgrade Watcher process to keep running so it can detect this repeated crashing or unhealthy status, and then initiate a rollback. 

By setting `KillMode=process` in Elastic Agent's `systemd` unit configuration file, we ensure that when the main process is repeatedly crashing, `systemd` keeps the Upgrade Watcher process running so it can fulfill it's destiny.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] ~I have added an integration test or an E2E test~

## How to test this PR locally

### Test A

For this test, we will upgrade from a pre-built version of Agent, say `8.9.0` to the build of Agent with this PR (`8.10.0`) and then to another build of Agent that deliberately crashes (say, `8.11.0`).

The upgrade from the first to the second build of Agent will test that the `/etc/systemd/system/elastic-agent.service` file starts off without the `KillMode=process` setting in it and that the setting is added after the new (post-upgrade) Agent starts running.

The upgrade from the second to the third build of Agent will test that the `KillMode=process` setting is behaving as expected, by ensuring that the Upgrade Watcher process keeps running even after the new (post-upgrade) Agent process crashes.

1. Start with a pre-built version of Elastic Agent < `8.10.0`
2. Unpack and install this Elastic Agent on Linux.
3. Check the contents of the `/etc/systemd/system/elastic-agent.service` file and ensure that the `KillMode=process` setting is absent.
   ```
   grep KillMode /etc/systemd/system/elastic-agent.service
   ```
4. Check out this PR's branch and build Elastic Agent for Linux.
   ```
   DEV=true PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
5. Upgrade to the newly-built Agent.
   ```
   sudo elastic-agent upgrade 8.10.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
   ```
6. Check the contents of the `/etc/systemd/system/elastic-agent.service` file and ensure that the `KillMode=process` setting is now present.  Don't worry about any whitespace around the `=` sign (`systemd` [ignores](https://www.freedesktop.org/software/systemd/man/systemd.syntax.html#) it).
   ```
   grep KillMode /etc/systemd/system/elastic-agent.service
   ```
7. Ensure that the `systemd` unit configuration files have been reloaded.  For this, run `systemctl status elastic-agent.service >/dev/null` and ensure that the following warning message is NOT shown:
   ```
   Warning: The unit file, source configuration file or drop-ins of elastic-agent.service changed on disk. Run 'systemctl daemon-reload' to reload units.
   ```

9. Ensure that the upgrade succeeds before moving on.  For this, wait until the Upgrade Watcher process has finished running and ensure that the running Elastic Agent's version is `8.10.0`.
   ```
   systemctl status elastic-agent.service | grep 'elastic-agent watch'
   sudo elastic-agent version
   ```
1. Bump up the minor version in `version/version.go` (to, say, `8.11.0`), and also return an error at the very start of https://github.com/elastic/elastic-agent/blob/af31709ed6d1d249235e2745eb3cc314d3a7f464/internal/pkg/agent/cmd/run.go#L65 to ensure that the built Agent will crash immediately upon start up.
2. Make sure to make a new commit with these changes so the build ID (SHA) in the Agent build is different. Otherwise the upgrade won't work.
4. Build Elastic Agent for Linux.
   ```
   DEV=true PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
   ```
5. Upgrade to the newly-built Agent.
   ```
   sudo elastic-agent upgrade 8.11.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
   ```
7. Monitor the output of `systemctl status elastic-agent.service`.  Watch the Upgrade Watcher process being spawned, then the new (post-upgrade) Agent process crashing, and ensure that the Upgrade Watcher process keeps running for at least 2 minutes.
   ```
   watch -n1 systemctl status elastic-agent.service
   ```

### Test B

For this test, we will upgrade from a pre-built version of Agent, say `8.9.0` to a build of Agent that has the changes in this PR but also crashes upon start up (say, `8.10.0`).

This will test that that the `/etc/systemd/system/elastic-agent.service` file starts off without the `KillMode=process` setting in it, that the setting is added after the new (post-upgrade) Agent starts running, and that the setting takes effect immediately so as to keep the Upgrade Watcher process running after the Agent process has crashed.

1. Start with a pre-built version of Elastic Agent < `8.10.0`
2. Unpack and install this Elastic Agent on Linux.
3. Check the contents of the `/etc/systemd/system/elastic-agent.service` file and ensure that the `KillMode=process` setting is absent.
   ```
   grep KillMode /etc/systemd/system/elastic-agent.service
   ```
4. Check out this PR's branch.
5. Modify the code to return an error right after the changes in this PR have a chance to take effect, to ensure that the built Agent will crash: https://github.com/elastic/elastic-agent/blob/af31709ed6d1d249235e2745eb3cc314d3a7f464/internal/pkg/agent/cmd/run.go#L108-L111

2. Make sure to make a new commit with these changes so the build ID (SHA) in the Agent build is different. Otherwise the upgrade won't work.
4. Build Elastic Agent for Linux.
   ```
   DEV=true PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
7. Upgrade to the newly-built Agent.
   ```
   sudo elastic-agent upgrade 8.10.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
   ```
8. Check the contents of the `/etc/systemd/system/elastic-agent.service` file and ensure that the `KillMode=process` setting is now present.  Don't worry about any whitespace around the `=` sign (`systemd` [ignores](https://www.freedesktop.org/software/systemd/man/systemd.syntax.html#) it).
   ```
   grep KillMode /etc/systemd/system/elastic-agent.service
   ```
9. Ensure that the `systemd` unit configuration files have been reloaded.  For this, run `systemctl status elastic-agent.service >/dev/null` and ensure that the following warning message is NOT shown:
   ```
   Warning: The unit file, source configuration file or drop-ins of elastic-agent.service changed on disk. Run 'systemctl daemon-reload' to reload units.
   ```
7. Monitor the output of `systemctl status elastic-agent.service`.  Watch the Upgrade Watcher process being spawned, then the new (post-upgrade) Agent process crashing, and ensure that the Upgrade Watcher process keeps running for at least 2 minutes.
   ```
   watch -n1 systemctl status elastic-agent.service
   ```

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Fixes #3123

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

elasticmachine commented:
## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.734% (`78/79`) | :+1: 
Files | 67.018% (`191/285`) | :+1: 
Classes | 65.472% (`347/530`) | :+1: 
Methods | 52.53% (`1090/2075`) | :+1: 
Lines | 38.376% (`12455/32455`) | :+1: 0.028
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

pchila reviewed (approved):
LGTM

ycombinator commented:
buildkite test this

blakerouse reviewed (changes requested):
I would like to talk about this change a little bit more.

I read all of the details from the issue that you provided on changing this value. I do worry that this change of the timeout does have a side-effect of `systemctl stop elastic-agent`. I believe that if stopping the Elastic Agent using the above command hangs that it would block for the full 10 minutes. I don't know if that is the behavior we want to provide in that situation.

I also worry about this having an effect on uninstall. We stop the service, if the watcher is running and uninstall is performed will it block for 10mins waiting for the watcher to die before being able to continue.

I actually thinking changing the `KillMode=process` is the proper solution. Being that Elastic Agent is a supervisor and it spawns all of it subprocess with the proper `Pkill` in the situations that require it. It would also equal the same behavior that we have on macOS and WIndows.

ycombinator commented:
Thanks for taking the time to read through all the investigation details in the issue and bringing up your concerns here. Let me do some more testing for the scenarios you outlined as well as with the alternative idea of specifying `KillMode=process` instead of the `TimeoutStopSec` setting and post my findings here.

ycombinator commented:
@blakerouse Thanks for taking the time to read through all the investigation details in the issue and bringing up your concerns here. Let me do some more testing for the scenarios you outlined as well as with the alternative idea of specifying `KillMode=process` instead of the `TimeoutStopSec` setting and post my findings here.

ycombinator commented:
I tested the three scenarios @blakerouse brought up in https://github.com/elastic/elastic-agent/pull/3220#pullrequestreview-1572182008.  Observations and conclusions from each of the three tests are below but tl;dr is that it seems better to go with the `KillMode=process` setting than the `TimeoutStopSec` setting.  I'll update this PR accordingly.

> I do worry that this change of the timeout does have a side-effect of `systemctl stop elastic-agent`. I believe that if stopping the Elastic Agent using the above command hangs that it would block for the full 10 minutes. I don't know if that is the behavior we want to provide in that situation.

I built a version of Agent (on top of the changes in this PR) that will block for 30 minutes (`<-time.After(30 * time.Minute)`) very early in the `elastic-agent run` code path.  

I installed this Agent made sure it started running under the service.

Then I immediately called `systemctl stop elastic-agent` and then looked at the output of `systemctl status elastic-agent`. 

It appears the service is stopped and the Agent process is terminated.

```
$ sudo systemctl status elastic-agent.service
Unit elastic-agent.service could not be found.
$ sudo ./elastic-agent install
Elastic Agent will be installed at /opt/Elastic/Agent and will run as a service. Do you want to continue? [Y/n]:y
Do you want to enroll this Agent into Fleet? [Y/n]:n
Elastic Agent has been successfully installed.
$ sudo systemctl status elastic-agent.service
● elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2023-08-10 19:50:35 UTC; 26s ago
   Main PID: 478576 (elastic-agent)
      Tasks: 5 (limit: 4637)
     Memory: 10.9M
        CPU: 60ms
     CGroup: /system.slice/elastic-agent.service
             └─478576 /opt/Elastic/Agent/elastic-agent

Aug 10 19:50:35 shaunak-ubuntu-22-arm systemd[1]: Started Elastic Agent is a unified agent to observe, monitor and protect your system..
$  sudo systemctl stop elastic-agent.service
$ sudo systemctl status elastic-agent.service
○ elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Thu 2023-08-10 19:51:11 UTC; 5s ago
    Process: 478576 ExecStart=/usr/bin/elastic-agent (code=killed, signal=TERM)
   Main PID: 478576 (code=killed, signal=TERM)
        CPU: 65ms

Aug 10 19:50:35 shaunak-ubuntu-22-arm systemd[1]: Started Elastic Agent is a unified agent to observe, monitor and protect your system..
Aug 10 19:51:11 shaunak-ubuntu-22-arm systemd[1]: Stopping Elastic Agent is a unified agent to observe, monitor and protect your system....
Aug 10 19:51:11 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Deactivated successfully.
Aug 10 19:51:11 shaunak-ubuntu-22-arm systemd[1]: Stopped Elastic Agent is a unified agent to observe, monitor and protect your system..
```

**Conclusion**: it would appear that setting `TimeoutStopSec=610` does not have an adverse impact on the behavior of `systemctl stop elastic-agent`.

---

> I also worry about this having an effect on uninstall. We stop the service, if the watcher is running and uninstall is performed will it block for 10mins waiting for the watcher to die before being able to continue.

I tested this scenario by starting with a build of Agent from this PR (so it includes the `TimeoutStopSec=610` setting) and upgrading to a build of Agent that crashes immediately upon starting up.

```
$ grep TimeoutStopSec /etc/systemd/system/elastic-agent.service
TimeoutStopSec=610
$ sudo elastic-agent upgrade 8.13.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
Upgrade triggered to version 8.13.0, Elastic Agent is currently restarting
$ 
```

As expected, due to the timeout change in this PR, the Upgrade Watcher process kept running after the upgraded Agent process crashed.

```
$ sudo systemctl status elastic-agent.service
● elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; enabled; vendor preset: enabled)
     Active: deactivating (stop-sigterm) (Result: exit-code) since Thu 2023-08-10 20:29:06 UTC; 3min 10s ago
    Process: 492383 ExecStart=/usr/bin/elastic-agent (code=exited, status=1/FAILURE)
   Main PID: 492383 (code=exited, status=1/FAILURE)
      Tasks: 6 (limit: 4637)
     Memory: 103.8M
        CPU: 6.234s
     CGroup: /system.slice/elastic-agent.service
             └─493106 /opt/Elastic/Agent/elastic-agent watch --path.config /opt/Elastic/Agent --path.home /opt/Elastic/Agent

Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --environment environmentVar   set environment being ran in (default default)
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:   -h, --help                         help for elastic-agent
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.config string           Config path is the directory Agent looks for its >
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.downloads string        Downloads path contains binaries Agent downloads >
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.home string             Agent root path (default "/opt/Elastic/Agent")
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.install string          Install path contains binaries Agent extracts
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.logs string             Logs path contains Agent log output (default "/op>
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:   -v, --v                            Log at INFO level
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]: Use "elastic-agent [command] --help" for more information about a command.
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]: prematurely crashing agent
```

At this point, I uninstalled the Agent. The command returned almost immediately, as usual, without blocking for too long.

```
$ sudo elastic-agent uninstall
Elastic Agent is installed but currently broken: service is not installed
Continuing will uninstall the broken Elastic Agent at /opt/Elastic/Agent. Do you want to continue? [Y/n]:y
Elastic Agent has been uninstalled.
```

Then I checked the service status.

```
$ sudo systemctl status elastic-agent.service
Warning: The unit file, source configuration file or drop-ins of elastic-agent.service changed on disk. Run 'systemctl daemon-reload' to reload uni>
● elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; bad; vendor preset: enabled)
     Active: deactivating (stop-sigterm) (Result: exit-code) since Thu 2023-08-10 20:29:06 UTC; 3min 31s ago
   Main PID: 492383 (code=exited, status=1/FAILURE)
      Tasks: 6 (limit: 4637)
     Memory: 103.4M
        CPU: 6.239s
     CGroup: /system.slice/elastic-agent.service
             └─493106 /opt/Elastic/Agent/elastic-agent watch --path.config /opt/Elastic/Agent --path.home /opt/Elastic/Agent

Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --environment environmentVar   set environment being ran in (default default)
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:   -h, --help                         help for elastic-agent
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.config string           Config path is the directory Agent looks for its >
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.downloads string        Downloads path contains binaries Agent downloads >
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.home string             Agent root path (default "/opt/Elastic/Agent")
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.install string          Install path contains binaries Agent extracts
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:       --path.logs string             Logs path contains Agent log output (default "/op>
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]:   -v, --v                            Log at INFO level
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]: Use "elastic-agent [command] --help" for more information about a command.
Aug 10 20:29:06 shaunak-ubuntu-22-arm elastic-agent[492383]: prematurely crashing agent
```

As the output above shows, the service was still running and, notably, the Upgrade Watcher process was still running. 

Interestingly, however, about 2m16s later, the Upgrade Watcher process stopped running, much sooner than the 660s timeout.  Given the odd timing (2m16s), most likely the Upgrade Watcher itself failed somehow.  Unfortunately since the `elastic-agent uninstall` had wiped out everything under `/opt/Elastic/Agent`, I couldn't inspect the Upgrade Watcher logs.  However, this does indicate to me, that had the Upgrade Watcher process not naturally exited it would likely have been kept running by `systemd` for 660s.  

Here we can see that the service is still running but there are no processes under it.

```
$ sudo systemctl status elastic-agent.service
Warning: The unit file, source configuration file or drop-ins of elastic-agent.service changed on disk. Run 'systemctl daemon-reload' to reload uni>
● elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; bad; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2023-08-10 20:35:21 UTC; 1s ago
   Main PID: 492383 (code=exited, status=1/FAILURE)
        CPU: 6.266s

Aug 10 20:35:21 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Failed with result 'exit-code'.
Aug 10 20:35:21 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Consumed 6.266s CPU time.
```

Exactly 2 minutes after the Upgrade Watcher process stopped running, the service itself got uninstalled.

```
$ sudo systemctl status elastic-agent.service
Unit elastic-agent.service could not be found.
```

**Conclusion**: it would appear that setting `TimeoutStopSec=610` does have some impact on the uninstall process but it does not block the uninstall process.

---

> I actually thinking changing the `KillMode=process` is the proper solution. Being that Elastic Agent is a supervisor and it spawns all of it subprocess with the proper `Pkill` in the situations that require it. It would also equal the same behavior that we have on macOS and WIndows.

I tested this scenario by starting with a build of Agent that set `KillMode=process` instead of `TimeoutStopSec=610` in the unit configuration file,  and upgrading to a build of Agent that crashes immediately upon starting up.

```
$ grep TimeoutStopSec /etc/systemd/system/elastic-agent.service
$ grep KillMode /etc/systemd/system/elastic-agent.service
KillMode=process
$ sudo elastic-agent upgrade 8.13.0 --source-uri file://$(pwd)/build/distributions/ --skip-verify
Upgrade triggered to version 8.13.0, Elastic Agent is currently restarting
```

As expected, due to `KillMode=process`, the Upgrade Watcher process kept running after the upgraded Agent process crashed.

```
$ sudo systemctl status elastic-agent.service
● elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2023-08-10 20:59:03 UTC; 1min 57s ago
    Process: 503266 ExecStart=/usr/bin/elastic-agent (code=exited, status=1/FAILURE)
   Main PID: 503266 (code=exited, status=1/FAILURE)
      Tasks: 6 (limit: 4637)
     Memory: 104.3M
        CPU: 6.068s
     CGroup: /system.slice/elastic-agent.service
             └─504544 /opt/Elastic/Agent/elastic-agent watch --path.config /opt/Elastic/Agent --path.home /opt/Elastic/Agent

Aug 10 20:59:03 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Main process exited, code=exited, status=1/FAILURE
Aug 10 20:59:03 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Failed with result 'exit-code'.
Aug 10 20:59:03 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Unit process 504544 (elastic-agent) remains running after unit stopped.
Aug 10 20:59:03 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Consumed 6.047s CPU time.
```

At this point, I uninstalled the Agent. The command returned almost immediately, as usual, without blocking for too long.

```
$ sudo elastic-agent uninstall
Elastic Agent is installed but currently broken: service is not installed
Continuing will uninstall the broken Elastic Agent at /opt/Elastic/Agent. Do you want to continue? [Y/n]:y
Elastic Agent has been uninstalled.
```

Then I checked the service status.

```
$ sudo systemctl status elastic-agent.service
Warning: The unit file, source configuration file or drop-ins of elastic-agent.service changed on disk. Run 'systemctl daemon-reload' to reload uni>
○ elastic-agent.service - Elastic Agent is a unified agent to observe, monitor and protect your system.
     Loaded: loaded (/etc/systemd/system/elastic-agent.service; bad; vendor preset: enabled)
     Active: inactive (dead) (Result: exit-code) since Thu 2023-08-10 21:05:33 UTC; 15s ago
   Main PID: 506788 (code=exited, status=1/FAILURE)
      Tasks: 6 (limit: 4637)
     Memory: 102.9M
        CPU: 5.377s
     CGroup: /system.slice/elastic-agent.service
             └─507047 /opt/Elastic/Agent/elastic-agent watch --path.config /opt/Elastic/Agent --path.home /opt/Elastic/Agent

Aug 10 21:05:20 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Main process exited, code=exited, status=1/FAILURE
Aug 10 21:05:20 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Failed with result 'exit-code'.
Aug 10 21:05:20 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Unit process 507047 (elastic-agent) remains running after unit stopped.
Aug 10 21:05:20 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Consumed 5.370s CPU time.
Aug 10 21:05:33 shaunak-ubuntu-22-arm systemd[1]: Stopped Elastic Agent is a unified agent to observe, monitor and protect your system..
Aug 10 21:05:33 shaunak-ubuntu-22-arm systemd[1]: elastic-agent.service: Consumed 5.370s CPU time.
```
As the output above shows, the service was still running and, notably, the Upgrade Watcher process was still running. 

I kept watching the output of `sudo systemctl status elastic-agent.service`.  Exactly 2 minutes after the Agent crash happened, the Upgrade Watcher process was killed AND the service was also uninstalled.  This is definitely a better experience than with using the `TimeoutStopSec` setting.  This behavior also alleviates a concern I had with using the `KillMode=process` setting: what if one of the Agent's component sub-processes, e.g. Metricbeat, hangs? Will it prevent the service from being uninstalled?  That doesn't appear to be the case, given what happened to the Upgrade Watcher process.

I also wanted to test that, with the `KillMode=process` setting a rollback would've still been executed had I not uninstalled the Agent while the Upgrade Watcher was running. I essentially repeated all the steps above, skipping over the Agent uninstall step.  Within a couple of minutes after the upgraded Agent crashed, the Upgrade Watcher initiated the rollback.  I confirmed this by checking the Upgrade Watcher logs.

*Conclusions*:
* Using `KillMode=process` does prevent `systemd` from prematurely killing off the Upgrade Watcher process, thus allowing it enough time to do its job.
* Using `KillMode=process`, when `elastic-agent uninstall` is executed while the Upgrade Watcher is running, the service (and any lingering processes, e.g. the Upgrade Watcher) are removed/killed faster than when using the  `TimeoutStopSec` setting.

pchila commented on internal/pkg/agent/install/svc.go:
```diff
@@ -106,3 +114,35 @@ const darwinLaunchdConfig = `<?xml version='1.0' encoding='UTF-8'?>
 </plist>
 `
+
+// A copy of the systemd config template from github.com/kardianos/service
+// with added .Config.Option.TimeoutStopSec option
```
Nit: this comment is still referring to the first fix...

```suggestion
// A copy of the systemd config template from github.com/kardianos/service
// with added .Config.Option.KillMode option
```

ycombinator replied:
Thanks for catching.  Fixed in ad152a41ef3ca16763d2f65cfea67b27e54864bd.

pchila commented on internal/pkg/agent/install/svc.go:
```diff
@@ -54,6 +54,14 @@ func newService(topPath string) (service.Service, error) {
 	}
 
+	if runtime.GOOS == "linux" {
+		// The github.com/kardianos/service library doesn't support TimeoutStopSec in their prebuilt template.
+		// This option allows to pass our own template for the systemd unit configuration, which is a copy
+		// of the prebuilt template with added TimeoutStopSec option
```
```suggestion
		// The github.com/kardianos/service library doesn't support KillMode in their prebuilt template.
		// This option allows to pass our own template for the systemd unit configuration, which is a copy
		// of the prebuilt template with added KillMode option
```

pchila reviewed (approved):
Just 2 small updates in the comments, looks good otherwise

pchila reviewed (approved):
LGTM

ycombinator commented:
buildkite test this

ycombinator commented:
> Should we update the systemd service file that is used for the DEB/RPM? Do we want it to act the same as when the `elastic-agent install` is used? Being that DEB/RPM cannot be self-upgraded it doesn't have to be the same, just didn't know if we want to ensure uniformity.

Yes, we should! @cmacknz had mentioned it earlier too, I think.  Thanks for the reminder.  I'll add those changes to this PR.

> How are we going to get this change in an upgrade? This will only be a fix for new installations.

Hmm, this is a tricky one.  In general I'm thinking we need some way to make this change during the upgrade process but I can already think of a few issues with that idea.  Let me experiment a bit and see what might be possible.  Do you think this issue is is a blocker for this PR?  Or could we address it in a follow up PR?

ycombinator commented on internal/pkg/agent/install/svc.go:
```diff
@@ -54,6 +54,14 @@ func newService(topPath string) (service.Service, error) {
+	if runtime.GOOS == "linux" {
+		// The github.com/kardianos/service library doesn't support KillMode in their prebuilt template.
+		// This option allows to pass our own template for the systemd unit configuration, which is a copy
+		// of the prebuilt template with added KillMode option
```
Added explanation in 4e731ddd27c53092b728404bcd03b88e88f77898.

ycombinator commented:
> Should we update the systemd service file that is used for the DEB/RPM? Do we want it to act the same as when the `elastic-agent install` is used? Being that DEB/RPM cannot be self-upgraded it doesn't have to be the same, just didn't know if we want to ensure uniformity.

Done in c3f934be8cd35c0ce732b066e55efd9ee7da0c6c.

blakerouse reviewed (approved):
@ycombinator Don't think we should block this PR on upgrades not updating this file. I do think it is something that we need to do though.

ycombinator commented:
> I can already think of a few issues with that idea [of updating the unit file during an upgrade]. 

The two issues I was/am thinking of are:

1. The `root` requirement, which @cmacknz also mentioned in https://github.com/elastic/elastic-agent/pull/3220#issuecomment-1675272578, especially as we're now actively working on removing this requirement for running Agent.  

   My preference here is to go ahead with this change now and deal with the `root` requirement when we deal with other aspects of Agent that require `root`.

2. Whether to update the unit configuration file (and issue the `systemd daemon-reload`) from the old (pre-upgrade) Agent as one of the upgrade steps OR to do this update as part of a "migration" in the new (post-upgrade) Agent.  There are pros/cons to each approach:
   * Making the change from the new Agent as part of starting up
      * Pro: users upgrading _to_ Agent v8.10.0 (assuming that's when the code is released) or later version will benefit.
      * Con: if the new Agent crashes immediately upon start, the unit configuration change will not get deployed.
   * Making the change from the old Agent as one of the upgrade steps
      * Pro: if the new Agent crashes immediately upon start, the unit configuration change will have been deployed.
      * Con: only users who upgrade _from_ Agent v8.10.0 or later  will benefit.

   My preference is to make the change from the new Agent as:
   - it fits with the philosophy of controlling more of the upgrade process from the new Agent, e.g. the change we recently made to launch the new Agent's watcher, as it gives us the ability to fix bugs with new releases.
   - it should be rare for an Agent to crash immediate upon start; hopefully this is the sort of thing we'd either be able to catch well before a release or, if it happens anyway, hopefully it would affect a small segment of users due to some corner case deployment configuration.

 WDYT?

cmacknz commented:
> My preference here is to go ahead with this change now and deal with the root requirement when we deal with other aspects of Agent that require root.

👍 Agreed.

> My preference is to make the change from the new Agent as:

Agreed. I would suggest thinking of this as yet another state convergence problem. The agent should eventually be able to edit its unit configuration file, and should periodically retry until the file matches what is expected. A single failure shouldn't cause the change not to take effect. Remember that the power from the system can be removed at essentially any moment, so anything we do that doesn't retry will fail in that scenario.

The only other problem to consider is if users have manually edited parts of the configuration file, which we can likely handle by only changing the keys we care about and not replacing the entire file.

ycombinator commented:
> Don't think we should block this PR on upgrades not updating this file. I do think it is something that we need to do though.

Based on recent comments on the implementation, I believe I can get this addition into this PR in time for the 8.10 FF.  So I'm going to do that instead of a follow up PR.

blakerouse commented:
With us updating the unit file on upgrade, I believe we are going to need to do one more test.

The open question I have in my head is what does systemd do when the KillMode is changed will the service is already running?

If systemd creates the cgroup at start time of the process it could be possible that systemd will still exhibit the same behavior until the service gets completely restarted. We need to determine if we need a completely true restart of the Elastic Agent after the reload for the KillMode change to fully take effect.

ycombinator commented:
> With us updating the unit file on upgrade, I believe we are going to need to do one more test.
> 
> The open question I have in my head is what does systemd do when the KillMode is changed will the service is already running?
> 
> If systemd creates the cgroup at start time of the process it could be possible that systemd will still exhibit the same behavior until the service gets completely restarted. We need to determine if we need a completely true restart of the Elastic Agent after the reload for the KillMode change to fully take effect.

Okay, I will test this out, thanks.

If it turns out we need to restart the `elastic-agent` service after the unit config change and `systemd daemon-reload`, that should be possible too with `systemctl restart elastic-agent` (or the equivalent Golang API call to the https://github.com/kardianos/service/ library).

ycombinator commented on internal/pkg/agent/cmd/svc_update.go:
```diff
@@ -0,0 +1,12 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
```
Should this file be moved to `internal/pkg/agent/application/upgrade/service_update.go` so it lives along side `internal/pkg/agent/application/upgrade/service.go`?

blakerouse replied:
Yes I think it should.

ycombinator commented on internal/pkg/agent/cmd/svc_update_linux.go:
```diff
@@ -0,0 +1,48 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
```
Should this file be moved to `internal/pkg/agent/application/upgrade/service_update_linux.go` so it lives along side `internal/pkg/agent/application/upgrade/service.go`?

blakerouse replied:
Yes.

ycombinator commented:
@cmacknz @blakerouse I've addressed both your recent feedback and updated this PR accordingly.  I've also updated the "How to test this PR locally" section with two tests, the second of which answers @blakerouse's open question from https://github.com/elastic/elastic-agent/pull/3220#issuecomment-1675393562 (the answer is that the `KillMode=process` setting takes effect immediately and has an impact on the _already running_ Agent service).

Please re-review when you get a chance.  Thanks!

blakerouse commented on internal/pkg/agent/cmd/svc_update_linux.go:
```diff
@@ -0,0 +1,61 @@
+	cfg.Section("Service").Key("KillMode").SetValue("process")
+	if err := cfg.SaveTo(unitFilePath); err != nil {
+		return fmt.Errorf("error writing updated systemd unit file [%s]: %w", unitFilePath, err)
+	}
```
I think we should NOTE that on a rollback this will not be changed back. The older Elastic Agent doesn't have a way to ensure that the service file is in its expected content.

I wonder how hard it would be to use the 3rd party `service` module that we use to render the file in memory and then just compare the contents of the file? That way we are not doing something specific here. It could just compare the 2 file contents and if different replace it with the new one and reload.

Is it not possible with the 3rd part `service` module to just have it overwrite the file instead of doing it here directly?

ycombinator replied:
> I wonder how hard it would be to use the 3rd party `service` module that we use to render the file in memory and then just compare the contents of the file? That way we are not doing something specific here. It could just compare the 2 file contents and if different replace it with the new one and reload.

I looked through https://pkg.go.dev/github.com/kardianos/service#section-documentation and I'm not seeing any APIs exposed by the `service` module that would help load the existing unit configuration file into memory (essentially a `service.Config` object) and compare it against our `service.Config` object.

> Is it not possible with the 3rd part `service` module to just have it overwrite the file instead of doing it here directly?

The reason I didn't go with the overwrite approach is because of a consideration that @cmacknz brought up in https://github.com/elastic/elastic-agent/pull/3220#issuecomment-1675311109: 

> The only other problem to consider is if users have manually edited parts of the configuration file, which we can likely handle by only changing the keys we care about and not replacing the entire file.

That's why I went with the approach of loading the current unit configuration file into memory, making a surgical edit if needed, and writing it back to disk.

blakerouse replied:
True, if they manually edited the file.

blakerouse commented on internal/pkg/agent/cmd/run.go:
```diff
@@ -579,6 +579,18 @@ func handleUpgrade() error {
+		return err
+	}
+
+	if err := ensureServiceConfigUpToDate(); err != nil {
```
I think it would be good to have this as `upgrade.EnsureServiceConfig()`. Related to your ask below.

ycombinator replied:
Done in 09ec4a120a90bc3e01d92db93c5887a9fc56401f.

blakerouse reviewed (commented):
@ycombinator Awesome work! Glad to hear that the `KillMode=process` take effect without a restart!

cmacknz reviewed (approved):
LGTM, thanks! I like how small the code to edit the systemd configuration turned out.

Could we modify one of the upgrade integration tests to check that KillMode process change takes effect? The best case would be to always do this on a test from a pre-8.10 version to 8.10+.

ycombinator commented:
> Could we modify one of the upgrade integration tests to check that KillMode process change takes effect? The best case would be to always do this on a test from a pre-8.10 version to 8.10+.

Sure, that will be done as part of https://github.com/elastic/elastic-agent/pull/3085. It was while writing the test in that PR that I found this bug.

blakerouse reviewed (changes requested):
See inline comment.

ycombinator commented on internal/pkg/agent/application/upgrade/service_update_linux.go:
```diff
@@ -0,0 +1,64 @@
+		}
+
+		// Reload systemd unit configuration files
+		cmd := exec.Command("systemctl", "daemon-reload")
```
Yeah, good point.  Will change!

ycombinator replied:
Done in beb7ce9273f67efce005130fa63075d2ade392b8.

blakerouse reviewed (approved):
Looks good! Thanks for addressing all the feedback.
